### ISSUE TITLE
Restarting from checkpoint does not work properly

### ISSUE THREAD

**Issue by vykhand** [dnouri/skorch#355]
Problem:

I am trying to restart training from a checkpoint, with a different loss criterion.
When I load model checkpoint like so:

```
_ = net.initialize()
net.load_params("../runs/EXP003_FOCAL/1/EXP003_FOCAL_154_best.params")
```
and run predict, it gives me the right result, which is the best previous IOU

```
val_preds = su.sigmoid(net.predict(val_ds))
me.iou_metric(val_masks, val_preds)
```

results in :

> 0.7820625

but when I restart training with partial_fit, or with warm_start = false, does not matter, the training starts from a much lower value and never gets back to the original value. 

```
 epoch    train_iou    train_loss    val_iou    valid_loss    cp       dur
-------  -----------  ------------  ---------  ------------  ----  --------
      1       0.7792        0.0644     *0.7425*        0.1492     +  262.0857
      2       0.7898        0.0608     *0.7291*        0.1409     +  263.2233
```

I have ran the same in plain vanilla torch, starting from the same checkpoint, and the training starts correctly from approx. IOU = 0.782. I've been looking into the skorch code and can't find the reason for such behaviour.

Apologies that I have not pinned this issue exactly to the line of code, but may be you will give me some ideas where to dig. I don't have a GPU locally so it is hard for me to run it in a debugger. Also it might be the case that I screwed things up myself and just don't see how.

**Comment by vykhand** [comment]
@githubnemo thank you! this is exactly the case, I am using Adam

**Comment by vykhand** [comment]
@githubnemo but still it is strange since I am comparing with plain vanilla torch, where I also re-init the optimizer from scratch and it starts from a much higher metric value.

**Comment by githubnemo** [comment]
Can you please post a minimal example that reproduces this behaviour?

**Comment by vykhand** [comment]
@githubnemo I am not sure, as I have a lot of custom code and it is for a kaggle competition, so can't post the whole thing publicly yet. I will first try to modify my forked code to save and restore all structures (scheduler, optimizer, etc.) and will see if it helps. Thanks again for giving me a good clue to investigate.

**Comment by BenjaminBossan** [comment]
> but when I restart training with partial_fit, or with warm_start = false, does not matter, the training starts from a much lower value and never gets back to the original value.

Just to be sure, this is lower value is still much better than at the start of the training? Otherwise, one possibility would be that the module is re-initialized for some reason.

**Comment by vykhand** [comment]
@BenjaminBossan yes, it is better than at the start of the training. The training starts with IoU close to ~0.6

**Comment by ottonemo** [comment]
It would be helpful if you could post the parameters you pass to the network constructor (and other parameters you set using, e.g. `set_params`).

Another thing to keep in mind is that the validation score is reported at the end of the epoch, so it may very well be that the skorch model continues training after loading the checkpoint with a val IOU of 0.78 but reports a much worse IOU at the end of the epoch due to overfitting.

**Comment by vykhand** [comment]
Hi Folks,

Reloading the from file indeed seems to solve the problem. Not sure why I was able to restart from a higher metric in Torch without restoring the optimizer state. I am closing the issue. Thanks again for your helpful comments.

**Comment by vykhand** [comment]
So my solution was to subclass the callback to save optimizer and add the restoring method in my model subclass:
```
class CheckpointWithOpt(Checkpoint):
    def __init__(self,
                 target=None,
                 monitor='valid_loss_best',
                 f_params='params.pt',
                 f_history=None,
                 f_pickle=None,
                 f_torch_struct=None):
        super(CheckpointWithOpt, self).__init__(target, monitor, f_params, f_history, f_pickle)
        self.f_torch_struct = f_torch_struct

    def save_model(self, net):
        super(CheckpointWithOpt, self).save_model(net)
        if self.f_torch_struct:
            torch_dict = {"optimizer": net.optimizer_}
            torch.save(torch_dict, self.f_torch_struct)

class SaltNeuralNet(NeuralNet):
.... skipped
    def load_torch_struct(self, f):
        torch_dict = torch.load(f)
        self.optimizer_ = torch_dict["optimizer"]
```

**Comment by benjamin-work** [comment]
@vykhand Glad that it worked and thanks for posting a solution. I created a new issue (#357) to add a feature in skorch to save the optimizer.

**Comment by vykhand** [comment]
@BenjaminBossan @benjamin-work I think in general saving the model should be somehow redesigned, because pickle does not work for me at all. I have not introduced any lambda functions, but I am still getting pickling errors when trying to serialize the NeuralNet with pickle. I can raise an issue, but I think model should save the important structures and not the whole thing like callbacks etc. which can be easily re-created.
